### PR TITLE
Enhance two integration tests

### PR DESCRIPTION
**Pull request by robinw0928**
Enhance test_append and test_array_creation
1.  add negative tests
2.  add more test cases
3.  refactor test code

### CODE REVIEW

File: tests/integration/utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    if len(b) != len(c):
+        is_equal = False
+        err_arr = [b, c]
+    else:
+        for each in zip(b, c):
+            if not np.array_equal(*each):
+                err_arr = each
+                is_equal = False
+                break
+    return is_equal, err_arr
```
Comment (bryevdv) [inline tests/integration/utils/utils.py:38]: It think this could be simplified:
```suggestion
    if len(b) != len(c):
        return False, [b, c]

    for each in zip(b, c):
        if not np.array_equal(*each):
            return False, each

    return True, None
```

Comment (bryevdv) [reply tests/integration/utils/utils.py:38]: I'm also just curious, is the point of this comparison to get better detail about which top-level sub-arrays are not equal?

Comment (robinw0928) [reply tests/integration/utils/utils.py]: I think so. I'm wondering why the function is called in many integration tests.

File: tests/integration/utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def run_test(fn, args, kwargs, print_msg):
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/utils/utils.py:35]: In my understanding, this method applies to every test which returns array as output. I find it a little hard to rename it for now. Can we call it run_test_with_array_output?

File: tests/integration/utils/utils.py (hunk 1 above)
Comment (bryevdv) [reply tests/integration/utils/utils.py]: I expect it is in order to obtain better diagnostic output in cases where there is a failure

File: tests/integration/test_array_creation.py [hunk 3]
```diff
@@ -181,25 +215,33 @@ def test_arange_with_dtype(args, dtype):
...
+
+    def test_zero_division(self):
+        with pytest.raises(ZeroDivisionError):
+            num.arange(0, 10, 0)
+        with pytest.raises(ZeroDivisionError):
+            num.arange(0.0, 10.0, 0.0)
+        with pytest.raises(ZeroDivisionError):
+            num.arange(0, 0, 0)
+        with pytest.raises(ZeroDivisionError):
+            num.arange(0.0, 0.0, 0.0)
```
Comment (bryevdv) [inline tests/integration/test_array_creation.py:242]: This seems like a good case to parametrize over tuples passed as `*args` especially since there are more cases that can be tested: nonzero (both positive and negative) start values should be covered explicitly

File: tests/integration/utils/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bryevdv) [reply tests/integration/utils/utils.py:35]: I would suggest something like `check_array_method` to reflect that it specifically compares outputs of arbitrary cunumeric/numpy array methods

File: tests/integration/test_append.py [hunk 4]
```diff
@@ -69,16 +38,46 @@ def _run_test(arr, values, test_args):
...
+        b = np.random.randint(low=0, high=100, size=size_b)
+        print_msg = f"np.append(array({a.shape}), array({b.shape}), {axis})"
+        run_test("append", [a, b], {"axis": axis}, print_msg)
+
+
+def test_append_axis_none():
+    axis = None
+    for size_a in SIZES:
+        a = np.random.randint(low=0, high=100, size=size_a)
+        for size_b in SIZES:
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_append.py]: Fixed.

File: tests/integration/test_append.py [hunk 5]
```diff
@@ -69,16 +38,46 @@ def _run_test(arr, values, test_args):
...
+    ):
+        num.append(a, b, axis=1)
+
+    size_c = (10, DIM)
+    c = np.random.randint(low=0, high=100, size=size_c)
+    with pytest.raises(AssertionError):
+        num.append(a, c, axis=1)
+
+    with pytest.raises(IndexError):
+        num.append(a, a, axis=5)
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_append.py]: @magnatelee It's found only in eager execution test, it raises ValueError. In other tests, it raises AssertionError.

File: tests/integration/test_array_creation.py [hunk 6]
```diff
@@ -32,80 +34,133 @@ def test_array():
...
-    yo = np.ones((2, 3))
-    assert np.array_equal(xo, yo)
-    assert xo.dtype == yo.dtype
+        xf = num_f(shape, dtype=dtype)
+        yf = np_f(shape, dtype=dtype)
 
+        if fn == "empty":
+            assert xf.shape == yf.shape
+        else:
+            assert np.array_equal(xf, yf)
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_array_creation.py]: Fixed. Thanks.

File: tests/integration/test_array_creation.py [hunk 7]
```diff
@@ -32,80 +34,133 @@ def test_array():
...
+        with pytest.raises(ValueError):
+            num.zeros(shape)
+        with pytest.raises(ValueError):
+            num.ones(shape)
+        with pytest.raises(ValueError):
+            num.full(shape, 10)
+
+    num.full((2, 3), [1])
+    with pytest.raises(AssertionError):
+        num.full((2, 3), [10, 20, 30])
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_array_creation.py:120]: It's found only in eager execution test, it raises ValueError. In other tests, it raises AssertionError.

File: tests/integration/test_array_creation.py [hunk 8]
```diff
@@ -32,80 +34,133 @@ def test_array():
...
+    np_f = getattr(np, fn)
+
+    x = num.array(x_np)
+    xfl = num_f(x, dtype=dtype)
+    yfl = np_f(x_np, dtype=dtype)
+
+    if fn == "empty_like":
+        assert xfl.shape == yfl.shape
+    else:
+        assert np.array_equal(xfl, yfl)
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_array_creation.py]: Fixed. Thanks.

File: tests/integration/test_array_creation.py [hunk 9]
```diff
@@ -117,13 +172,36 @@ def test_arange(args):
...
+        num.arange(0, 1, num.nan)
+
+    with pytest.raises(ZeroDivisionError):
+        num.arange(0, 10, 0)
+    with pytest.raises(ZeroDivisionError):
+        num.arange(0.0, 10.0, 0.0)
+    with pytest.raises(ZeroDivisionError):
+        num.arange(0, 0, 0)
+    with pytest.raises(ZeroDivisionError):
+        num.arange(0.0, 0.0, 0.0)
```
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/test_array_creation.py]: Fixed. Thanks.

File: tests/integration/utils/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (robinw0928) [reply tests/integration/utils/utils.py]: Fixed. Thanks.

File: tests/integration/test_array_creation.py (hunk 3 above)
Comment (robinw0928) [reply tests/integration/test_array_creation.py]: Fixed. Thanks.

File: tests/integration/test_array_creation.py [hunk 10]
```diff
@@ -214,13 +214,27 @@ def test_arange_with_dtype(args, dtype):
...
+    (1.0, 10.0, 0.0),
+    (10.0, -10.0, 0.0),
+]
+
+
 class TestArrangeErrors:
     def test_negative_sizes(self):
         with pytest.raises(ValueError):
-            ###np.arange(-10) returns [] successfully
+            # np.arange(-10) returns [] successfully
```
Comment (robinw0928) [inline tests/integration/test_array_creation.py:234]: @magnatelee Is this difference in results between numpy and cunumeric here expected or a bug?

File: tests/integration/test_array_creation.py [hunk 11]
```diff
@@ -189,8 +189,8 @@ def test_full_like_assertion():
     (3.5,),
     (2, 10),
     (-2.5, 10.0),
-    ### output: num: array([ 1, -1, -3, -5, -7]),
-    ### np: array([ 1. , -1.5, -4. , -6.5, -9. ]
+    # output: num: array([ 1, -1, -3, -5, -7]),
+    # np: array([ 1. , -1.5, -4. , -6.5, -9. ]
     # (1, -10, -2.5),
```
Comment (robinw0928) [inline tests/integration/test_array_creation.py:194]: @magnatelee Is this difference in results between numpy and cunumeric here expected or a bug?

File: tests/integration/test_array_creation.py [hunk 12]
```diff
@@ -76,24 +85,38 @@ def test_full(value):
 SHAPES_NEGATIVE = [
     -1,
     (-1, 2, 3),
-    # num.array([2, -3, 4]),  ## it would raise RuntimeError("Unable to find attachment to remove") when num.array
-    # is removed at the end as global variable
+    ## it raises RuntimeError("Unable to find attachment to remove")
```
Comment (robinw0928) [inline tests/integration/test_array_creation.py]: @magnatelee Is this expected or an issue here? Thanks

File: tests/integration/test_array_creation.py (hunk 10 above)
Comment (bryevdv) [reply tests/integration/test_array_creation.py:234]: note that in case there are currently expected failures, we can still have the test run (and avoid leaving commented code in the repo) by using `@pytest.mark.xfail`

**Comment by manopapad** [comment]
AFAICT all the differences that @robinw0928 has pointed out above are real bugs, including the cases where cunumeric asserts-out instead of throwing a `ValueError` like NumPy does. @robinw0928 can you open issues for these, so we can track work on fixing them? Alternatively, feel free to submit fixes within this PR if it's something small.

File: tests/integration/test_array_creation.py (hunk 10 above)
Comment (manopapad) [reply tests/integration/test_array_creation.py:234]: I like this approach. If an `xfail` test switches to succeeding (when we fix the corresponding bug), will that show up as a test suite failure?

Comment (bryevdv) [reply tests/integration/test_array_creation.py:234]: @manopapad yes it will which is a nice way to know in case you've accidentally fixed a bug :)

Comment (robinw0928) [reply tests/integration/test_array_creation.py:234]: @bryevdv I like this approach, too. For AssertionError raised by cuNumeric, do we need to expect ValueError as NumPy in  our test code and make it xfail?

Comment (bryevdv) [reply tests/integration/test_array_creation.py:234]: >  do we need to expect ValueError as NumPy in our test code and make it xfail?

That is certainly an option although if this is a very simple fix I'd say just fix it (to raise ValueError) in this PR if possible.

Comment (robinw0928) [reply tests/integration/test_array_creation.py]: AssertionError is fixed. It raises ValueError now. Other failures are added with xfail.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
LGTM!

File: cunumeric/deferred.py [hunk 13]
```diff
@@ -708,7 +708,11 @@ def _broadcast(self, shape: NdShape) -> Any:
 
         for dim in range(len(shape)):
             if result.shape[dim] != shape[dim]:
-                assert result.shape[dim] == 1
+                if result.shape[dim] != 1:
+                    raise ValueError(
+                        f"Shape did not match along dimension {dim}"
```
Comment (manopapad) [inline cunumeric/deferred.py:713]: Leave a space at the end of the string here, otherwise when they get joined it will look like this: "... dimension 3and the ..."

Comment (robinw0928) [reply cunumeric/deferred.py]: Fixed. Thanks. @manopapad After this review, can I merge the change?

**Review by marcinz — APPROVED** [review]
